src/main/java/beam/physsim/cch/RoutingFrameworkTravelTimeCalculator.scala

```diff
@@ -0,0 +1,230 @@
+
+    logger.info("Finished creation of graph {}", System.currentTimeMillis - startTime)
+
+    val id2Link = links.toStream.map(x => x.getId.toString.toInt -> x).toMap
+
+    val hour2Events: Map[Int, List[TravelInfo]] = generateHour2Events(pathTraversalEvents)
+
+    val odsCreationFutures = hour2Events
+      .map {
+        case (hour, events) =>
+          Future {
+            val stopWatch: StopWatch = new StopWatch
+            stopWatch.start()
+
+            var odNumber = 0
+
+            val ods = events.toStream
+              .map { event =>
+                for {
+                  firstLinkId <- event.linkIds.headOption
+                  lastLinkId  <- event.linkIds.lastOption
+                  firstWayId  <- linkWayId(id2Link(firstLinkId))
+                  secondWayId <- linkWayId(id2Link(lastLinkId))
+                  origin      <- osmInfoHolder.getCoordinatesForWayId(firstWayId).headOption
+                  destination <- osmInfoHolder.getCoordinatesForWayId(secondWayId).lastOption
+                } yield origin -> destination
+              }
+              .collect { case Some(od @ (origin, destination)) if origin != destination => od }
+              .map {
+                case (origin, destination) =>
+                  val firstVertexId: Long = coord2VertexId
+                    .getOrElse(origin, getClosestVertexId(origin))
+                  val secondVertexId: Long = coord2VertexId
+                    .getOrElse(destination, getClosestVertexId(destination))
+
+                  odNumber = odNumber + 1
+
+                  OD(firstVertexId, secondVertexId)
+              }
+
+            val odStream = ods
+              .flatMap(od => Stream.range(0, odsFactor, 1).map(_ => od))
+            routingFrameworkWrapper.generateOd(iterationNumber, hour, odStream)
+
+            logger.info("Generated {} ods, for hour {} in {} ms", odNumber, hour, stopWatch.getTime)
+          }
+      }
+
+    Await.result(Future.sequence(odsCreationFutures), 10.minutes)
+
+    val hour2Way2TravelTimes = hour2Events.keys.toList.sorted.map { hour =>
+      val stopWatch: StopWatch = new StopWatch
+      stopWatch.start()
+
+      logger.info("Starting traffic assignment for hour {}", hour)
+      val way2TravelTime =
+        routingFrameworkWrapper.assignTrafficAndFetchWay2TravelTime(iterationNumber, hour)
+      logger.info("Traffic assignment for hour {} is finished in {} ms", hour, stopWatch.getTime)
+
+      (hour, way2TravelTime)
+    }.toMap
+
+    val totalNumberOfLinks: Int = links.size
+    var linksFailedToResolve = 0
+
+    val linkId2TravelTimeByHour = links
+      .groupBy(linkWayId)
+      .flatMap {
+        case (maybeWayId, linksInWay) =>
+          linksInWay.map { link =>
+            val travelTimeByHour = maybeWayId match {
+              case Some(wayId) =>
+                var speedFoundForAtLeastOneHour = false
+
+                val linkTravelTimeByHour = Array.tabulate(maxHour) { hour =>
```

Comment:
Maybe this can be extracted to the function so one can write unit tests for at least this small parts?